pkg/declcfg/write.go

```diff
@@ -0,0 +1,161 @@
+package declcfg
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/operator-framework/operator-registry/pkg/property"
+)
+
+const (
+	globalName = "__global"
+)
+
+func WriteDir(cfg DeclarativeConfig, configDir string) error {
+	entries, err := ioutil.ReadDir(configDir)
+	if err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	if len(entries) > 0 {
+		return fmt.Errorf("config dir %q must be empty", configDir)
+	}
+
+	return writeToFS(cfg, &diskWriter{}, configDir)
+}
+
+type fsWriter interface {
+	MkdirAll(path string, mode os.FileMode) error
+	WriteFile(path string, data []byte, mode os.FileMode) error
+}
+
+var _ fsWriter = &diskWriter{}
+
+type diskWriter struct{}
+
+func (w diskWriter) MkdirAll(path string, mode os.FileMode) error {
+	return os.MkdirAll(path, mode)
+}
+
+func (w diskWriter) WriteFile(path string, data []byte, mode os.FileMode) error {
+	return ioutil.WriteFile(path, data, mode)
+}
+
+func writeToFS(cfg DeclarativeConfig, w fsWriter, rootDir string) error {
+	bundlesByPackage := map[string][]Bundle{}
+	for _, b := range cfg.Bundles {
+		bundlesByPackage[b.Package] = append(bundlesByPackage[b.Package], b)
+	}
+	othersByPackage := map[string][]Meta{}
+	for _, o := range cfg.Others {
+		pkgName := o.Package
+		if pkgName == "" {
+			pkgName = globalName
```

Comment:
I can see this being useful, but I'm also wondering if we just want to split all blobs into groups: entries that pass the validation for their schema and entries that don't. I imagine an error message saying "package X isn't valid: missing name" is more useful than seeing the `__global` folder with invalid data in it?

Reply:
This piece of code is not concerned with validation at all. It is just splitting blobs into package-specific vs non-package specific so that it can write blobs out in places convenient for humans to later find them.

I do think it could be interesting to add more logic to this to further split out blobs into valid/invalid, but I think that's orthogonal. For example, you might want `__global-invalid.json` and `<package>/<package>-invalid.json`